### PR TITLE
dxTreeList -  Implementation of the forEachNode method

### PR DESCRIPTION
**Pull request by Alyar666**

### CODE REVIEW

**Review by TikhonovaTatyana — DISMISSED** [review]

File: js/ui/tree_list/ui.tree_list.data_controller.js [hunk 1]
```diff
@@ -224,6 +224,16 @@ exports.DataController = dataControllerModule.controllers.data.inherit((function
             }
 
             return this._dataSource.loadDescendants(keys, childrenOnly);
+        },
+
+        /**
+         * @name dxTreeListMethods_forEachNode
+         * @publicName forEachNode(nodes, function)
+         * @param1 nodes:Array<dxTreeListNode>
+         * @param2 function:function
```
Comment (lukyanovas) [inline js/ui/tree_list/ui.tree_list.data_controller.js]: Maybe `callback ` name will be more clear here

Comment (Alyar666) [reply js/ui/tree_list/ui.tree_list.data_controller.js]: @TikhonovaTatyana @RomanTsukanov ?

Comment (IlyaKhD) [reply js/ui/tree_list/ui.tree_list.data_controller.js]: `dxTreeListMethods_callback` seems to be too uncertain. `nodeProcessingCallback` or smth like ?

**Review by IlyaKhD — APPROVED** [review]

File: js/ui/tree_list/ui.tree_list.data_controller.js (hunk 1 above)
Comment (RomanTsukanov) [reply js/ui/tree_list/ui.tree_list.data_controller.js]: Assuming that this is any JavaScript function at all, why should it be called anything other?

**Review by RomanTsukanov — DISMISSED** [review]

**Review by lukyanovas — CHANGES_REQUESTED** [review]
Please, just rename @param2 function:function somehow. 'function' is not a good name for the parameter.

File: js/ui/tree_list/ui.tree_list.data_controller.js (hunk 1 above)
Comment (IlyaKhD) [reply js/ui/tree_list/ui.tree_list.data_controller.js]: Sorry, I've mixed up a bit, didn't follow you.
`callback` is the best name suggested, no need for any more clarification. Moreover, it's a "_traditional_" name for such params, I think.

P.S. `function` could be treated as a reserved word in this case,

**Review by IlyaKhD — CHANGES_REQUESTED** [review]

File: js/ui/tree_list/ui.tree_list.data_controller.js (hunk 1 above)
Comment (TikhonovaTatyana) [reply js/ui/tree_list/ui.tree_list.data_controller.js]: If `function` does not suit
I think `callback` is a good variant

File: js/ui/tree_list/ui.tree_list.data_source_adapter.js [hunk 2]
```diff
@@ -613,6 +613,13 @@ DataSourceAdapter = DataSourceAdapter.inherit((function() {
                 .fail(d.reject);
 
             return d.promise();
+        },
+
+        forEachNode: function(nodes, callback) {
+            if(typeUtils.isObject(nodes)) {
```
Comment (vconst) [inline js/ui/tree_list/ui.tree_list.data_source_adapter.js]: Change isObject condition to Array.isArray

**Review by IlyaKhD — APPROVED** [review]

**Review by lukyanovas — APPROVED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**